src/index.js

```diff
@@ -349,126 +271,228 @@ export default class RestClient {
     }
 
-    if (errors) return { errors, status: res.status };
-    let body = await res[options.streamReader]();
-    body = options.bodyParser(body, context);
-    if (body.errors) return { errors: body.errors, status: res.status };
-    return { data: body.data, headers: res.headers, status: res.status };
+    if (errors) {
+      errors = castArray(errors);
+      return this._resolveFetch(context, options, res, { endpoint, errors });
+    }
+
+    const { ok, status, statusText } = res;
+    const responseGroup = getResponseGroup(status);
+    const _metadata = { ...metadata, endpoint, method, ok, responseGroup, status, statusText };
+
+    if (responseGroup === 'redirection' && res.headers.get('location')) {
+      return this._handleRedirect(method, fetchOptions, context, options, res, _metadata);
+    }
+
+    if (responseGroup === 'serverError') {
+      return this._handleRetry(
+        method, endpoint, fetchOptions, context, options, res, _metadata,
+      );
+    }
+
+    return this._resolveFetch(context, options, res, _metadata);
   }
 
   /**
    *
    * @private
    * @param {string} endpoint
-   * @param {string|Array<string>} values
+   * @param {Array<string>} values
    * @param {Object} context
    * @param {Object} options
    * @param {string} resourceKey
    * @return {Promise}
    */
   async _get(endpoint, values, context, options, resourceKey) {
-    const res = await this._fetch('GET', endpoint, context, options);
-    const data = res.data ? castArray(res.data) : [];
-    const errors = res.errors;
-    const _values = values ? castArray(values) : [];
-    this._resolveRequests({ data, errors }, _values, context, resourceKey);
-    if (errors) return errors;
-
-    if (data.length) {
-      this._setCache(data, res.headers, context.path, context.queryParams, resourceKey);
+    const cache = await this._checkCache(endpoint);
+    const { cacheability } = cache;
+    let res, fromCache;
+
+    if (!cacheability || cacheability.noCache || !cacheability.check()) {
+      const headers = { ...options.headers };
+      const etag = cacheability && cacheability.etag;
+      if (etag) headers['If-None-Match'] = etag;
+      const method = 'GET';
+      const fetchOptions = { headers: new Headers(headers), method };
+      const metadata = { errors: [], resource: values };
+      res = await this._fetch(method, endpoint, fetchOptions, context, options, metadata);
+    } else {
+      fromCache = true;
+      const requests = this._getRequestsTracker(context.path, context.method, context.fetchID);
+      requests.push({ endpoint, fromCache, resource: values });
     }
 
-    if (!context.resource) return data;
-    return this._resolveResource(data, resourceKey, context.resource.active);
+    const metadata = res ? res.metadata : {};
+    const data = fromCache || metadata.status === 304 ? cache.data : res.data;
+    if (metadata.ok || metadata.status === 304) this._setCacheEntry(endpoint, data, res.headers);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') TypeError can be thrown as "res" might be null or undefined here. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2259)